#### By the end of today, we should understand:  

* What a tokenizer does
* Why LLMs use tokens instead of words
* Special tokens ([CLS], [SEP], [PAD], etc.)
* input_ids
* attention_mask
* How text becomes numbers
This is one of the most important concepts in LLM Engineering.  

#### Why Can't Models Understand Text?   
Suppose we give BERT:  

**I love Machine Learning**
  
Humans understand it immediately.   

Computers only understand:   

0, 1, 2, 3, ...  

So we need a `translator`:  

Text  
 ↓  
Tokenizer  
 ↓   
Numbers   
 ↓    
Transformer Model   

The `tokenizer` is that `translator`.

### Let's Inspect BERT's Tokenizer

In [1]:
from transformers import AutoTokenizer

In [2]:
tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")

text = "I am learning LLM engineering."

tokens = tokenizer.tokenize(text)
print(tokens)

['i', 'am', 'learning', 'll', '##m', 'engineering', '.']


**Notice:**   
ll   
##m   

BERT didn't know "LLM" as a complete word.   

So it split it into smaller pieces.    

`This is called:`   
Subword Tokenization

### Why Subword Tokenization?   
Imagine a vocabulary containing:   
cat  
dog   
house   
book  
`Now we encounter:`   
ChatGPT   
What should the model do?   
`Instead of failing:`   
ChatGPT   
`becomes`   
Chat   
##GP   
##T   
or similar.   
`This allows models to understand:`   
New words   
Misspellings   
Rare words   
`without retraining.`

### Convert Text to IDs

In [3]:
encoded = tokenizer("I am learning LLM engineering")
print(encoded)

{'input_ids': [101, 1045, 2572, 4083, 2222, 2213, 3330, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1]}


`input_ids`   
These are token numbers.    
as we have 'input_ids': [101, 1045, 2572, 4083, 2222, 2213, 3330, 102]   
`Think:`    
Word → Token → ID   
`Example:`    
learning → 4083  

### Decode IDs Back

In [4]:
print(encoded["input_ids"])

[101, 1045, 2572, 4083, 2222, 2213, 3330, 102]


In [5]:
decoded = tokenizer.decode(encoded["input_ids"])

In [6]:
print(decoded)

[CLS] i am learning llm engineering [SEP]


### Special Tokens   
[CLS] and [SEP]   

* start of sentence [CLS].
* end of sentence [SEP].

### Attention Mask

In [7]:
print(encoded['attention_mask'])

[1, 1, 1, 1, 1, 1, 1, 1]


##### means:
1 = Pay attention   
0 = Ignore

#### Exercise 1

In [8]:
words = [
    "ChatGPT",
    "Artificial Intelligence",
    "MachineLearningEngineer",
    "Transformers"
]

for word in words:
    print(word)
    print(tokenizer.tokenize(word))
    print("-"*50)

ChatGPT
['chat', '##gp', '##t']
--------------------------------------------------
Artificial Intelligence
['artificial', 'intelligence']
--------------------------------------------------
MachineLearningEngineer
['machine', '##lea', '##rn', '##ingen', '##gin', '##eer']
--------------------------------------------------
Transformers
['transformers']
--------------------------------------------------


### Exercise 2

In [9]:
sentence1 = "I learn AI"
sentence2 = "I absolutely learn Artificial Intelligence"

print(len(tokenizer.tokenize(sentence1)))
print(len(tokenizer.tokenize(sentence2)))

3
5


sentence1 have 3 tokens   
sentence2 have 5 tokens

What is tokenization?    
Tokenization is the process of breaking text into smaller units called tokens so that a language model can process and understand the text.

Why do LLMs use tokens instead of words?   
LLMs use tokens because tokens can represent complete words, parts of words, punctuation, and special symbols. This helps the model handle unknown words, rare words, and multiple languages more efficiently.

What are special tokens?  
Special tokens are predefined tokens with specific purposes, such as marking the start, end, or padding of a sequence. Examples include [CLS], [SEP], and [PAD].

What are input_ids?   
input_ids are the numerical representations of tokens. Since neural networks work with numbers, each token is converted into a unique ID before being fed into the model.

What is attention_mask?   
attention_mask tells the model which tokens should be considered (1) and which tokens should be ignored (0), especially when padding is used.